projects/libsodium/fake_random.h

```diff
@@ -0,0 +1,46 @@
+#ifndef FAKE_RANDOM_H_
+#define FAKE_RANDOM_H_
+
+#include <assert.h>
+#include <sodium.h>
+#include <stdint.h>
+
+#include "fake_random.h"
+
+static const char *
+fake_implementation_name(void) {
+  return "fake_random";
+}
+
+static uint32_t
+fake_randombytes(void) {
+  return 0;
+}
+
+static void
+fake_random_buffer(void * const buf, const size_t size) {
+  static const unsigned char constant_seed[randombytes_SEEDBYTES] = {   \
+    'N', 'O', 'T', 'E', 'S', 'T', 'D', 'A', 'T', 'A',                   \
+    'N', 'O', 'T', 'E', 'S', 'T', 'D', 'A', 'T', 'A',                   \
+    'N', 'O', 'T', 'E', 'S', 'T', 'D', 'A', 'T', 'A',                   \
+    'N', '0'};
+  randombytes_buf_deterministic(buf, size, constant_seed);
+}
+
+struct randombytes_implementation fake_random = {
+  .implementation_name = fake_implementation_name,
+  .random = fake_randombytes,
+  .stir = NULL,
+  .uniform = NULL,
+  .buf = fake_random_buffer,
+  .close = NULL
+};
+
+void
+setup_sodium_w_deterministic_random() {
+  assert(randombytes_set_implementation(&fake_random) == 0);
```

Comment:
Hmm...maybe my request for assertions was misguided.

My concern now is that these function calls may not be included in a binary that is compiled with `-DNDEBUG`.

I assume that these fuzzers are not built with this flag but I'm not sure, have you built the fuzzers and run them?

If the fuzzers were built with `NDEBUG` by accident at a later date, would they crash immediately or fail silently?

Ideally the result of something like `sodium_init` would be saved and the assertion performed on the variable storing its value. But if you think this current style is safe then its up to you.

Reply:
I've not compiled using `NDEBUG, but have built and run the fuzzers succesfully. Your point about the asserts being compiled away makes complete sense (I hadn't thought about this happening) - if the asserts get compiled away then so could the initialization code.